### DISCUSSION TITLE
Sionna for Basic Channel Estimation

### DISCUSSION THREAD

**Discussion by kmlfrkbsrn** [NVlabs/sionna#614] [General]
How to use Sionna for Basic Channel Estimation? I have already checked OFDM channel model. I'm trying use in 4X4 MIMO  LDPC channel, but I couldn't do this. Is it possible?

**Comment by SebastianCa** [comment]
Hi @kmlfrkbsrn,
please have a look at the [OFDM MIMO detection tutorial](https://nvlabs.github.io/sionna/examples/OFDM_MIMO_Detection.html). This should explain the basics of channel estimation in Sionna.

**Comment by kmlfrkbsrn** [comment]
Hi @SebastianCa , 
Thnx for reply. I have already check this page. But I couldn't adapt this to basic comm. channel.

**Reply by SebastianCa** [reply]
What type of channel do you use? Could you please provide small code snippet that explains where you are conceptually stuck?

**Reply by kmlfrkbsrn** [reply]
`sionna.config.xla_compat=True
class Model(tf.keras.Model):
    def __init__(self, num_tx_ant, num_rx_ant, spatial_corr=None):
        super().__init__()
        self.n = 1024
        self.k = 512
        self.coderate = self.k / self.n
        self.num_bits_per_symbol = 4
        self.num_tx_ant = num_tx_ant
        self.num_rx_ant = num_rx_ant
        self.binary_source = BinarySource()
        self.encoder = LDPC5GEncoder(self.k, self.n)
        self.mapper = Mapper("qam", self.num_bits_per_symbol)
        self.demapper = Demapper("app", "qam", self.num_bits_per_symbol)
        self.decoder = LDPC5GDecoder(self.encoder, hard_out=True)
        self.channel = FlatFadingChannel(self.num_tx_ant,
                                         self.num_rx_ant,
                                         spatial_corr=spatial_corr,
                                         add_awgn=True,
                                         return_channel=True)

    
   '@tf.function(jit_compile=True)'
    def call(self, batch_size, ebno_db):
        b = self.binary_source([batch_size, self.num_tx_ant, self.k])
        c = self.encoder(b)
        x = self.mapper(c)

        shape = tf.shape(x)
        x = tf.reshape(x, [-1, self.num_tx_ant]) 
        
        no = ebnodb2no(ebno_db, self.num_bits_per_symbol, self.coderate)
        no *= np.sqrt(self.num_rx_ant)

        y, h = self.channel([x, no])

        #Channel Estimation
        pilot_symbols = tf.constant([[1.0 + 1j, 1.0 - 1j, 1.0 + 1j, 1.0 - 1j]], dtype=tf.complex64)  # 1x4 pilot symbols

        y_pilot = y[:, :1]  # Size [batch_size, num_rx_ant]

        y_pilot_real = tf.math.real(y_pilot)
        y_pilot_imag = tf.math.imag(y_pilot)

        pilot_symbols_real = tf.math.real(pilot_symbols)
        pilot_symbols_imag = tf.math.imag(pilot_symbols)

        pilot_symbols_real_t = tf.transpose(pilot_symbols_real)
        pilot_symbols_imag_t = tf.transpose(pilot_symbols_imag)

        h_est_real = tf.linalg.matmul(y_pilot_real, tf.linalg.pinv(pilot_symbols_real_t))
        h_est_imag = tf.linalg.matmul(y_pilot_imag, tf.linalg.pinv(pilot_symbols_imag_t))

        h_est = tf.complex(h_est_real, h_est_imag)

        # LMMSE Equalizer
        s = tf.complex(no*tf.eye(self.num_rx_ant, self.num_rx_ant), 0.0)
        x_hat, no_eff = lmmse_equalizer(y, h_est, s)

        num_symbols = shape[0] // self.num_tx_ant 
        new_shape = tf.stack([batch_size, num_symbols, self.num_tx_ant])

        x_hat = tf.reshape(x_hat, new_shape)
        no_eff = tf.reshape(no_eff, new_shape)

        llr = self.demapper([x_hat, no_eff])

        llr = tf.reshape(llr, [batch_size, self.num_tx_ant, self.n])  

        b_hat = self.decoder(llr)

        return b, b_hat, h, h_est 
`
This is my channel code and estimation testing.

**Reply by SebastianCa** [reply]
I think there is something conceptually wrong with your code. For channel estimation you typically rely on known pilots (i.e., they are part of the transmitted signal `x` at known positions). See [LS estimator](https://nvlabs.github.io/sionna/api/ofdm.html#lschannelestimator) for a detailed explanation. In your case, you send random payload bits - and, thus, random symbols- but you define the value of the pilots only at the receiver. 

A simple sanity check of your code is to assume that the receiver knows a fraction of the transmitted symbols and use these know symbols for channel estimation.

**Reply by kmlfrkbsrn** [reply]
`pilot_symbols = tf.constant([[1.0 + 1j, 1.0 - 1j, 1.0 + 1j, 1.0 - 1j]], dtype=tf.complex64)  # 1x4 pilot symbols`

In here I gave the pilot symbol for testing channel. I know the output bits. I have curiosity about Is channel estimation technics available in Sionna only for OFDM channel?

**Reply by kmlfrkbsrn** [reply]
In addition, Are there any way to use below code for other channels estimations? Such as LDPC or AWGN channel.
  
`sionna.mimo.lmmse_equalizer(y, h, s, whiten_interference=True)`

**Reply by SebastianCa** [reply]
> `pilot_symbols = tf.constant([[1.0 + 1j, 1.0 - 1j, 1.0 + 1j, 1.0 - 1j]], dtype=tf.complex64) # 1x4 pilot symbols`
> 
> In here I gave the pilot symbol for testing channel. I know the output bits. I have curiosity about Is channel estimation technics available in Sionna only for OFDM channel?

The idea of channel estimation is to compare the received signal `y` against the known transmitted pilots `x_p`. Assuming ``y=h*x_p+n`` where `n` is some additional noise term, the channel estimator can then estimate `h`.  See [LS estimator](https://nvlabs.github.io/sionna/api/ofdm.html#lschannelestimator) for a more detailed explanation.

In the code above, you define the pilots only at the receiver. However, the transmitter sends random payload bits (and thus, random QAM symbols). Thus, there is a mismatch between the pilots you provide to the channel estimator and the actual transmitted symbols. Without these anchor points, the used channel estimator does not work.

**Reply by SebastianCa** [reply]
> In addition, Are there any way to use below code for other channels estimations? Such as LDPC or AWGN channel.
> 
> `sionna.mimo.lmmse_equalizer(y, h, s, whiten_interference=True)`

This is for MIMO channels where the system has multiple transmit/receive antennas. Please see [OFDM MIMO channel estimation tutorial](https://nvlabs.github.io/sionna/examples/OFDM_MIMO_Detection.html) for further details.

Note that for an AWGN channel no equalizer is required as the channel only adds Gaussian noise.
